src/ray/core_worker/core_worker.cc

```diff
@@ -127,4 +166,157 @@ std::unique_ptr<worker::ProfileEvent> CoreWorker::CreateProfileEvent(
       new worker::ProfileEvent(profiler_, event_type));
 }
 
+void CoreWorker::SetCurrentTaskId(const TaskID &task_id) {
+  worker_context_.SetCurrentTaskId(task_id);
+  // Clear all actor handles for non-actor tasks.
+  if (actor_id_.IsNil()) {
+    for (const auto &handle : actor_handles_) {
+      RAY_CHECK_OK(gcs_client_->Actors().AsyncUnsubscribe(handle.first, nullptr));
+    }
+    actor_handles_.clear();
+  }
+}
+
+TaskID CoreWorker::GetCallerId() const {
+  TaskID caller_id = GetCurrentTaskId();
+  ActorID actor_id = GetActorId();
+  if (!actor_id.IsNil()) {
+    caller_id = TaskID::ForActorCreationTask(actor_id);
+  }
+  return caller_id;
+}
+
+bool CoreWorker::AddActorHandle(std::unique_ptr<ActorHandle> actor_handle) {
+  const auto &actor_id = actor_handle->GetActorID();
+  auto inserted = actor_handles_.emplace(actor_id, std::move(actor_handle)).second;
+  if (inserted) {
+    // Register a callback to handle actor notifications.
+    auto actor_notification_callback = [this](const ActorID &actor_id,
+                                              const gcs::ActorTableData &actor_data) {
+      if (actor_data.state() == gcs::ActorTableData::RECONSTRUCTING) {
+        // We have to reset the actor handle since the next instance of the
+        // actor will not have the last sequence number that we sent.
+        auto it = actor_handles_.find(actor_id);
+        RAY_CHECK(it != actor_handles_.end());
+        it->second->Reset();
```

Comment:
Hmm, what happens to those tasks submitted before this reset happens (i.e. is a race possible?)

Reply:
Are you talking about the counters for those tasks? The task specs have already been created so it should be okay.

Reply:
Yeah, I'm thinking of the case when a task spec gets created with a old sequence number, and gets sent to the new actor? What will happen, will it hang or something?

Reply:
Ah I see. Right now, the direct actor transport only fails pending requests for actors that are `DEAD` instead of actors that are `RECONSTRUCTING`. I think it would work if we just failed pending requests in both cases.